### ISSUE TITLE
Ordering of WasmPlugin and AuthorizationPolicy

### ISSUE THREAD

**Issue by amar1104** [istio/istio#45511]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

Hi folks,
I have an AuthorizationPolicy with custom action which connects to external authz. I’m trying to add a WasmPlugin before AuthorizationPolicy but seems like it is always coming next to external AuthorizationPolicy. I have tried with AUTHN and AUTHZ phase in WasmPlugin and with high priority but still no luck.

SampleCode
```
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: auth-wasm
  namespace: default
spec:
  selector:
    matchLabels:
      app: nginx
  url: http://host.minikube.internal:9000/main.wasm
  phase: AUTHN
  priority: 100
```
```
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: auth-policy
  namespace: default #Change
spec:
  selector:
    matchLabels:
      app: nginx  #Change
  action: CUSTOM
  provider:
    name: test-ext-authz-http
  rules: 
    - {}
```
Expecting WasmPlugin to come after AuthPolicy but not working.

### Version

```Text
istioctl version: 1.17.2
kubectl version: 1.27.2
```

### Affected product area

- [ ] Ambient
- [ ] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Upgrade
- [ ] Multi Cluster
- [ ] Virtual Machine
- [ ] Control Plane Revisions

**Comment by hzxuzhonghu** [comment]
Can you share your listener configuration, from https://github.com/istio/istio/blob/8652668835700c4e278dac4437a9e736602bc559/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L363-L383, we do order them as documented

**Comment by amar1104** [comment]
@hzxuzhonghu sharing the listeners config here 
[envoy-config.txt](https://github.com/istio/istio/files/11788466/envoy-config.txt)

In the above code snippet I see this
`// TODO: how to deal with ext-authz? It will be in the ordering twice`
`filters = append(filters, lb.authzCustomBuilder.BuildHTTP(httpOpts.class)...) `
`filters = extension.PopAppend(filters, wasm, extensions.PluginPhase_AUTHN) `
Does it mean that ext-authz(custom AuthzPolicy) is added before PluginPhase_AUTHN?

**Comment by hzxuzhonghu** [comment]
IC, you are using custom authz, it is inserted in front of all others.

@howardjohn Do you remember why you put it like this in #38595

**Comment by howardjohn** [comment]
I think that PR was just retaining the existing ordering which was probably accidental

**Comment by hzxuzhonghu** [comment]
@yangminzhu Tracked custom authz filter was put in front by you, do you remember the ctx?

**Comment by amar1104** [comment]
Hi @yangminzhu, any update on this?

**Comment by howardjohn** [comment]
Yangmin is no longer working on the project so I wouldn't expect a response

**Comment by hzxuzhonghu** [comment]
Just not sure if there was some more considerations

**Comment by boeboe** [comment]
How about we define a new PluginPhase_AUTHZ_CUSTOM (as this is reflecting current reality) and modify protobuf accordingly, so customers can also inject custom WASM plugins before this new stage (which is now not possible).


https://github.com/istio/istio/blob/7c506844cf3f58b504c97a1a2e5a41ca0a37d25d/pilot/pkg/networking/core/v1alpha3/listener_builder.go#L396

```golang
		// TODO: how to deal with ext-authz? It will be in the ordering twice
		filters = extension.PopAppendHTTP(filters, wasm, extensions.PluginPhase_AUTHZ_CUSTOM)
		filters = append(filters, lb.authzCustomBuilder.BuildHTTP(httpOpts.class)...)
		filters = extension.PopAppendHTTP(filters, wasm, extensions.PluginPhase_AUTHN)
		filters = append(filters, lb.authnBuilder.BuildHTTP(httpOpts.class)...)
		filters = extension.PopAppendHTTP(filters, wasm, extensions.PluginPhase_AUTHZ)
		filters = append(filters, lb.authzBuilder.BuildHTTP(httpOpts.class)...)
		// TODO: these feel like the wrong place to insert, but this retains backwards compatibility with the original implementation
		filters = extension.PopAppendHTTP(filters, wasm, extensions.PluginPhase_STATS)
		filters = extension.PopAppendHTTP(filters, wasm, extensions.PluginPhase_UNSPECIFIED_PHASE)
```

https://github.com/istio/api/blob/master/extensions/v1alpha1/wasm.pb.go#L291

```
// The phase in the filter chain where the plugin will be injected.
type PluginPhase int32

const (
	// Control plane decides where to insert the plugin. This will generally
	// be at the end of the filter chain, right before the Router.
	// Do not specify `PluginPhase` if the plugin is independent of others.
	PluginPhase_UNSPECIFIED_PHASE PluginPhase = 0
	// Insert plugin before Istio authentication filters.
	PluginPhase_AUTHZ_CUSTOM PluginPhase = 1
	// Insert plugin before Istio authentication filters.
	PluginPhase_AUTHN PluginPhase = 2
	// Insert plugin before Istio authorization filters and after Istio authentication filters.
	PluginPhase_AUTHZ PluginPhase = 3
	// Insert plugin before Istio stats filters and after Istio authorization filters.
	PluginPhase_STATS PluginPhase = 4
)

// Enum value maps for PluginPhase.
var (
	PluginPhase_name = map[int32]string{
		0: "UNSPECIFIED_PHASE",
		1: "AUTHZ_CUSTOM",
		2: "AUTHN",
		3: "AUTHZ",
		4: "STATS",
	}
	PluginPhase_value = map[string]int32{
		"UNSPECIFIED_PHASE": 0,
		"AUTHZ_CUSTOM":      1,
		"AUTHN":             2,
		"AUTHZ":             3,
		"STATS":             4,
	}
)

```

This would require a change in the protobuf API definition, so I do not know what other impact his has. 

Original sample code of this issue adjusted:

```yaml
apiVersion: extensions.istio.io/v1alpha1
kind: WasmPlugin
metadata:
  name: auth-wasm
  namespace: default
spec:
  selector:
    matchLabels:
      app: nginx
  url: http://host.minikube.internal:9000/main.wasm
  phase: AUTHZ_CUSTOM
  priority: 100
---
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: auth-policy
  namespace: default #Change
spec:
  selector:
    matchLabels:
      app: nginx  #Change
  action: CUSTOM
  provider:
    name: test-ext-authz-http
  rules: 
    - {}
```

**Comment by kyessenov** [comment]
Strong -1 on adding new phases before we sort out the existing order.

**Comment by boeboe** [comment]
The ask of the market is clearly flexibility. 

Your remark that any AuthZ plugin (even a custom one) should always run after an AuthN plugin because to requires to pass data, does not make any sense. Custom plugins, specifically those designed on WASM will often combine AuthN and AuthZ.

My proposal is flexible and has no impact on existing cases.

**Comment by kyessenov** [comment]
You can solve this by adding new phases to Wasm, or you can solve it by changing where ext_authz is placed in the chain. Besides, the order of HTTP MX, cors, etc are also significant. My recommendation is to rationalize the extension ordering in a consistent way, that is not tied to the current inadequate Istio implementation. AUTHZ_CUSTOM is an Istio quirk, it is not something that's portable to other possible mesh implementations.

**Comment by kyessenov** [comment]
CC @keithmattix who's also adding Lua to the mix.

**Comment by keithmattix** [comment]
>  My recommendation is to rationalize the extension ordering in a consistent way, that is not tied to the current inadequate Istio implementation

I agree with this, but is it practical for us to do in Istio give backwards-compatibility challenges?

**Comment by kyessenov** [comment]
Having a documented ordering of extensions that is not in code would be a good progress, for example. Note that Istio is still changing order version-to-version, we did not make any promise on the fixed ordering between versions yet.

**Comment by zhaohuabing** [comment]
I believe the current implementation is not in line with the Istio documentation. According to the Istio documentation below, we should put the WASM plugin before custom authz.


Name | Description
-- | --
UNSPECIFIED_PHASE | Control plane decides where to insert the plugin. This will generally be at the end of the filter chain, right before the Router. Do not specify PluginPhase if the plugin is independent of others.
AUTHN | Insert plugin before Istio authentication filters.
AUTHZ | Insert plugin before Istio authorization filters and after Istio authentication filters.
STATS | Insert plugin before Istio stats filters and after Istio authorization filters.

**Comment by kyessenov** [comment]
Right, before we add any more complexity about ordering, we need to clean the house and get some internal consistency to start with. I'd like to see a complete ordering, authn/authz/telemetry/MX/Envoyfilter/Lua/Wasm/alpn/cors/rate limit all sorted correctly. @boeboe I understand that you want wasm before authz, but do you need ext_authz to be first? 

CC @aryan16

**Comment by kyessenov** [comment]
Also, it's very plausible to have two instances of the same filter in two places. I am aware of some ext_authz uses that need it in two places.

**Comment by boeboe** [comment]
@kyessenov the use case at hand is a wasm filter that needs to preceed the custom ExtAuthZ filter to do predecoration.

I agree the final solution needs cleanup and a lot more flexibility, but let's take small steps.

**Comment by howardjohn** [comment]
Every day we see requests to add 1 more small hack to address 1 more small
use case. In isolation each one is fine - if we accept 1 a day then the
project will fall apart under it's own complexity (if it's not too late
already).
Definitely agree with Kuat that we need a better understanding of the
ordering

**Comment by linsun** [comment]
I agree, we need to look at the requests holistically. Is someone willing to collect all Wasm ordering requests and come up with a proposal?

**Comment by howardjohn** [comment]
Related: #46899

**Comment by kyessenov** [comment]
@linsun I can try to organize that information, and we can discuss later how to evolve Istio to be more internally consistent.

**Comment by eduardosanzb** [comment]
Thanks a lot for the update and bring this issue back!
Is there any chance to re-evaluate the decision of the order for the `custom-authz` filter? 

Depending on the angle when reading the documentation, could seem a bit inconsistent having some kind of _authz_ running before the authn _(even tho it seems this has been like this forever)_; Given that authz relies on identity; there could be a hidden dependency from `authz` on `authn`.

Please take this comment just as a food for thought; and thanks again for your contributions.

**Comment by linsun** [comment]
> @linsun I can try to organize that information, and we can discuss later how to evolve Istio to be more internally consistent.

Perfect, thank you @kyessenov - I'd love to help as needed.

**Comment by keithmattix** [comment]
I am also happy to help if more hands are needed :)

**Comment by hanxiaop** [comment]
not stale

**Comment by linsun** [comment]
not stale

**Comment by sschepens** [comment]
I created a new issue to tracke the ordering of CUSTOM AuthorizationPolicy: https://github.com/istio/istio/issues/51173

**Comment by MaYuan-02** [comment]
> I think that PR was just retaining the existing ordering which was probably accidental

@howardjohn  Could we change the order of the HTTP filter chain to position JWT authentication before external authorization ? This doesn't seem to harm anyone. I would like to utilize the JWT information in the external authorization service, and it feels somewhat strange for JWT authentication to occur after JWT authorization.

**Comment by hanseltime** [comment]
Any news on this?  I get that there are probably a lot of existing workflows that we would have to be careful to not break, but I would say that having an AUTHZ before an AUTHN is a bug in the conceptual contract.  From my standpoint, I have waypoints with WASM plugins that I would like to replicate on my ingresses so I can avoid additional network hops. 

These WASM plugins are doing header sanitization that HAS to be done before my ext_authz runs (separation of concerns).  Right now, I have to maintain a lua filter and the WASM plugin together or setup ingress -> waypoint routing, which your own defaults caution against.

**As a solution,** I would propose that we make some sort of opt-in reconfiguration or additional phase.

If we agreed this was an underlying failure in the AUTHZ model, we would have a flag like V2_AUTH_PHASES and then when that is enabled, we would either reorder ext_AUTHZ (but I think that's maybe not the most correct) or we would add a new phase `ext_authn` and `ext_authz` would be placed with the authz phases.  Then wasm_authn would happen in a consistent order among authn's and likewise for authz.  People wanting to make the migration could do a clear:

1. Set the istiod flag
2. Update all ext_authz phases that were actually authn to the new ext_authn
